From f002c4d1d6785f4d65d2f839eb047af8841dc531 Mon Sep 17 00:00:00 2001
From: Gurpreet Ghai <gghai@codeaurora.org>
Date: Thu, 30 Jan 2020 11:32:32 +0530
Subject: [PATCH 22/37] BT_Audio: Updated Check for TWS+ switch

- Race condition between disconnection of first earbud and
  connection of second earbud, the configuration of second
  earbud is ignored because it is considered a case of TWS
  switch.
- Since MM Audio removes the disconnected device from list
  after delay of ~600ms, this creates a window where a new
  connection goes unprocessed when it happens immediately
  after disconnection.
- This change fetches real time connection state from A2DP
  to overcome the race condition.

CRs-Fixed: 2606382

Change-Id: I2df8447acb1d8e1b843a49906f2f3e3993cce5d6
---
 .../core/java/com/android/server/audio/BtHelper.java  | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/audio/BtHelper.java b/services/core/java/com/android/server/audio/BtHelper.java
index c360980f584..ba526310013 100644
--- a/services/core/java/com/android/server/audio/BtHelper.java
+++ b/services/core/java/com/android/server/audio/BtHelper.java
@@ -71,7 +71,7 @@ public class BtHelper {
     private @Nullable BluetoothHearingAid mHearingAid;
 
     // Reference to BluetoothA2dp to query for AbsoluteVolume.
-    private @Nullable BluetoothA2dp mA2dp;
+    static private @Nullable BluetoothA2dp mA2dp;
 
     // If absolute volume is supported in AVRCP device
     private boolean mAvrcpAbsVolSupported = false;
@@ -222,13 +222,18 @@ public class BtHelper {
     /*packages*/ @NonNull static boolean isTwsPlusSwitch(@NonNull BluetoothDevice device,
                                                                  String address) {
         BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
-        if (device == null || adapter.getRemoteDevice(address) == null ||
+        BluetoothDevice connDevice = adapter.getRemoteDevice(address);
+        if (device == null || connDevice == null ||
             device.getTwsPlusPeerAddress() == null) {
             return false;
         }
         if (device.isTwsPlusDevice() &&
-            adapter.getRemoteDevice(address).isTwsPlusDevice() &&
+            connDevice.isTwsPlusDevice() &&
             device.getTwsPlusPeerAddress().equals(address)) {
+            if(mA2dp.getConnectionState(connDevice) != BluetoothProfile.STATE_CONNECTED) {
+                Log.w(TAG,"Active earbud is already disconnected");
+                return false;
+            }
             Log.i(TAG,"isTwsPlusSwitch true");
             return true;
          }
-- 
2.28.0

